compiler/hash-typecheck/src/ops/validate.rs

```diff
@@ -88,110 +298,388 @@ impl<'gs, 'ls, 'cd> Validator<'gs, 'ls, 'cd> {
-                // definition attached.
-                enum MergeKind {
-                    Unknown,
-                    _Level2,
-                    _Level1 { nominal_attached: Option<TermId> },
-                }
-                let _merge_kind = MergeKind::Unknown;
-                for term_id in terms.iter().copied() {
-                    let reader = self.reader();
-                    let term = reader.get_term(term_id);
-                    match term {
-                        Term::Level2(_) => todo!(),
-                        Term::Access(_) => todo!(),
-                        Term::Var(_) => todo!(),
-                        Term::Merge(_) => todo!(),
-                        Term::TyFn(_) => todo!(),
-                        Term::TyFnTy(_) => todo!(),
-                        Term::AppTyFn(_) => todo!(),
-                        Term::AppSub(_) => todo!(),
-                        Term::Unresolved(_) => todo!(),
-                        Term::Level3(_) => todo!(),
-                        Term::Level1(_) => todo!(),
-                        Term::Level0(_) => todo!(),
-                        _ => return Err(TcError::InvalidElementOfMerge { term: term_id }),
-                    }
+                // Validate the level of each term against the merge restrictions (see
+                // [Self::validate_merge_element] docs).
+                let mut merge_kind = MergeKind::Unknown;
+                for merge_element_term_id in terms.iter().copied() {
+                    self.validate_merge_element(
+                        &mut merge_kind,
+                        simplified_term_id,
+                        merge_element_term_id,
+                    )?;
                 }
 
                 Ok(result)
             }
-            Term::TyFn(_) => {
-                // Validate each member.
-                // Ensure the return value of each case is a subtype of the general return type.
-                todo!()
-            }
+
+            // Level 1 terms:
             Term::Level1(level1_term) => match level1_term {
-                Level1Term::Tuple(_) => {
+                Level1Term::Tuple(tuple_ty) => {
                     // Validate each parameter
-                    todo!()
+                    let tuple_ty = tuple_ty.clone();
+                    self.validate_params(&tuple_ty.members)?;
+
+                    // Ensure each parameter is runtime instantiable:
+                    for param in tuple_ty.members.positional() {
+                        self.ensure_term_is_runtime_instantiable(param.ty)?;
+                    }
+                    Ok(result)
                 }
-                Level1Term::Fn(_) => {
+                Level1Term::Fn(fn_ty) => {
                     // Validate parameters and return type
-                    todo!()
+                    let fn_ty = fn_ty.clone();
+                    self.validate_params(&fn_ty.params)?;
+                    self.validate_term(fn_ty.return_ty)?;
```

Comment:
Perhaps, but it is hard to reason about which errors can be "plugged" to continue typechecking and which need to terminate. For now all operations terminate on the first error (other than when it obviously shouldn't be the case). I think we should deal with this in separate PRs, focusing on error reporting, because this consideration needs to be taken in many places in the typechecker (and it is not easy to know when unless we first have typechecked Hash code). What do you think?

Reply:
This sounds like a good approach, let’s focus on functionality first before fully considering user experience